[GSoC'24] added documentation for consensus feature (#8401)
cvat-ai/cvat: site/content/en/docs/manual/advanced/consensus.md

> ---

Reviewer: Could you check the file via a spell checker?

title: 'Consensus'
linkTitle: 'Consensus'
weight: 26
description: 'Consensus based annotation'
---

By involving multiple annotators, consensus-based annotation incorporates a range of perspectives and interpretations, which helps in minimising individual biases that might skew the annotations.

Multiple annotators reviewing and agreeing on annotations can help identify and correct errors or inconsistencies that might arise from individual biases.

For example creating bounding boxes around the face of a Basset Hound dog. One person might consider the ears also as the part of dogs face and one might not, this is due to fact that different people would have different understanding of things around them. In such cases, consensus based approach helps in reducing this bias by considering the interpretation through consensus.

| | |
|---|---|

> |![basket_hound_ear_covered](/images/basket_hound_ear_covered.jpeg)|![basket_hound_ear_not_covered](/images/basket_hound_ear_not_covered.jpeg)|

Reviewer: Probably, it's basset, not basket.

Author: Uh Yes, sorry my bad

- [Step by Step Guide](#step-by-step-guide)

## Use Cases

The basic use case is annotating the entire dataset with multiple annotators to obtain high-quality annotations in the dataset. In some cases it may work, but typically, however, this way of annotating is prohibitively expensive, as you need to annotate everything several times. There are several ways how the situation can be improved.

1. You can use several highly skilled annotators and rely on their consensus to create high-quality Ground Truth annotations. Typically, the Ground Truth subset contains a small portion of the whole dataset (e.g. 2-5%), so it's relatively easy to annotate it several times. These annotations are used then to validate work of different annotators randomly, so it's very desirable to avoid single expert bias in Ground Truth, and this is one of the key benefits of the consensus annotation. This method requires extra work, but it's much lower than annotating all the dataset several times. For instance, annotating a 3% of images 5 times will result in extra 3% * (5 - 1) = 12% overhead, comparing to single GT annotations.

> 2. In cases of points and poly-line / poly-shape annotations, annotators can add more annotations without a fixed structure, which can introduce significant variation. Similarly, for skeleton annotations, while there is a structure, annotators can move the points around. Due to the presence of numerous points, this can lead to significant variations in annotations for the same data. Using a consensus-based approach for such data types can help reduce these variations, which may arise from minor errors in placing the points.

Reviewer: Actually, this use case is not covered completely yet.

Reviewer: As we're only using one of the input shapes as a whole and have not implemented per annotation element merging.

Author: Nice, makes sense

Author: So, would it better to remove this use case, or let it be here as aparaently that's a use case for this feature. Though in a toned down sense?

> 2. In cases of points and poly-line / poly-shape annotations, annotators can add more annotations without a fixed structure, which can introduce significant variation. Similarly, for skeleton annotations, while there is a structure, annotators can move the points around. Due to the presence of numerous points, this can lead to significant variations in annotations for the same data. Using a consensus-based approach for such data types can help reduce these variations, which may arise from minor errors in placing the points.
>    1. For example, in the figure, an annotator might place a point representing the foot anywhere across the pink region, leading to significant variation in the annotations. ![pink_shoes](/images/pink_shoes.jpeg)

Reviewer: ```suggestion
2. In cases of points and poly-line / poly-shape annotations, annotators can add more annotations without a fixed structure, which can introduce significant variation. Similarly, for skeleton annotations, while there is a structure, annotators can move the points around. Due to the presence of numerous points, this can lead to significant variations in annotations for the same data. Using a consensus-based approach for such data types can help reduce these variations, which may arise from minor errors in placing the points.  For example, in the figure, an annotator might place a point representing the foot anywhere across the pink region, leading to significant variation in the annotations. Using such annotations for model training of for model validation is likely to result in reduced performance scores, exactly because of these variations in the correct annotations of the same objects. The consensus annotation approach allows you to minimize such kind of errors by using averaged annotations instead.

![pink_shoes](/images/pink_shoes.jpeg)
```

## Terminology:

> 1. `min_iou_threshold`: This is the same as `iou_threshold` param in Quality Settings. It's used for distinction between matched / unmatched annotations. A annotation is considered in the consensus only with the annotations it has an overlap greater than or equal to this

Reviewer: ```suggestion
1. `min_iou_threshold`: This is the same as `iou_threshold` param in Quality Settings. It's used for distinction between matched / unmatched annotations. An annotation is considered in the consensus only with the annotations it has an overlap greater than or equal to this
```

> 2. `agreement_score_threshold`: After consensus the merged annotation, is assigned a score based on it's overlap with other annotations in the cluster, (add label score thing also). Merged annotation with this score lesser than `agreement_score_threshold` are discarded.

Reviewer: ```suggestion
2. `agreement_score_threshold`: After consensus has produced a merged annotation, it is assigned a score based on it's overlap with other annotations in the cluster, (add label score thing also). Merged annotations with this score lesser than `agreement_score_threshold` are discarded.
```

> 2. `agreement_score_threshold`: After consensus the merged annotation, is assigned a score based on it's overlap with other annotations in the cluster, (add label score thing also). Merged annotation with this score lesser than `agreement_score_threshold` are discarded.

Reviewer: Will you add `(add label score thing also)`?

Author: Should I add it as here in a cluster all the annotations will be of the same label?

Reviewer: Just make the resulting values understandable to the users. If you're sure the label score is always 1, it can be removed.

> 3. `Quorum`: The minimum number of annotations in a consensus for the merging to occur. While deciding the final label of merged annotation, the count of specific label shouldn't be less than this.

Reviewer: ```suggestion
3. `Quorum`: The minimum number of annotations in a consensus for the merging to occur. While deciding the final label of a merged annotation, the count of specific label shouldn't be less than this. Clusters having less than this number of annotations are discarded.
```

   2. ![frame 2](/images/worker_2_2.jpeg) In this frame worker 2 hasn't annotated one of the dogs.
   3. ![frame 3](/images/worker_2_3.jpeg)
6. Click on `Merge Consensus Jobs`, this will merge all the consensus jobs into their regular job.
   1. ![consensus merge button](/images/consensus_merge_button.jpeg)
   2. ![consensus merge confirmation](/images/consensus_merge_confirmation.jpeg)
   3. ![loading consensus merge](/images/loading_consensus_merge.jpeg) The button will be disabled until the previous merge action isn't completed.
7. Below are the merged results
   1. ![frame 1](/images/merged_1.jpeg)
   2. ![frame 2](/images/merged_2.jpeg) Through consensus it got the annotation from worker 1's annotation
   3. ![frame 3](/images/merged_3.jpeg)
8. On clicking on `View Consensus Analytics` ![view consensus analytics](/images/view_consensus_analytics.jpeg)
   1. ![consensus analytics](/images/consensus_analytics_1.jpeg) Here the score corresponding to the job (merged job) represents, the mean overlap (IoU) among the annotations in it's consensus jobs and the merged job.

>    2. ![consensus analytics conflicts](/images/consensus_analytics_1_conf.jpeg) There was an annotation made by worker 1 but not by worker 2, that annotation didn't had any matching annotations so it raised `No matching annotation` conflict. Similarly, there were two annotations in the first frame which didn't matched as they didn't satisfied the `Min Overlap Threshold` condition. Thus, 3 `No matching annotation` conflicts are raised.

Reviewer: I think the example is a little bit hard to gauge, unless the quorum is set to 2. Basically, there is no other consensus measurement can be between only 2 sources. I suggest either to increase the number or sources to 3 or add more examples.

2. `agreement_score_threshold`: After consensus has produced a merged annotation, it is assigned a score based on it's
   overlap with other annotations in the cluster, (add label score thing also). Merged annotations with this score
   lesser than `agreement_score_threshold` are discarded.
3. `Quorum`: The minimum number of annotations in a consensus for the merging to occur. While deciding the final label
   of a merged annotation, the count of specific label shouldn't be less than this. Clusters having less than this
   number of annotations are discarded.
4. `oks_sigma`: This is the same as `oks_sigma` param in Quality Settings. Like `min_iou_threshold`, but for points.
   The percent of the bbox area, used as the radius of the circle around the GT point, where the checked point is
   expected to be. Read more: [https://cocodataset.org/#keypoints-eval](https://cocodataset.org/#keypoints-eval)
5. `line_thickness`: This is the same as `line_thickness` param in Quality Settings. Thickness of polylines, relatively
   to the (image area) ^ 0.5. The distance to the boundary around the GT line, inside which the checked line points
   should be.

> ## Working

Reviewer 2: A bit strange name for the section. It describes the algorithm details. Maybe `Algorithm Description`, `Workflow` or `Feature Details` would be better? What do you think?

Author: So true, I didn't notice much. `Algorithm Description` sounds good, thanks ^^

   5. In this, we keep track of the annotations we have added to a cluster. To avoid going through the matching
      annotation multiple times, We are doing so as we can come across an annotation multiple times since an annotation
      can be a matching annotation to multiple annotations from other consensus jobs
3. Merging cluster into single annotation:
   1. A cluster having annotations less than quorum is discarded
   2. Create a bounding box corresponding to every annotation
   3. Calculate the mean bounding box for these bounding boxes
   4. Go through the bounding box representation of the annotations in the cluster, and choose the annotation which is
      closest to the mean bounding box as the merged annotation
   5. Steps 3 and 4 are done differently for skeleton-type annotation, where the merged annotation is the one which has
      the least mean distance with other annotations in cluster
   6. The consensus score of this merged annotation is the mean of the overlap between the merged annotation and the
      rest of the annotations in cluster
   7. The merged annotation with a consensus score less than the agreement score threshold is discarded.

> ## Finding Overlap between two annotations:

Reviewer 2: ```suggestion
## Finding Overlap between two annotations
```

2. Points type: Object Keypoint Similarity (OKS)
3. PolyLine type:
   1. We first derive another polyline having the same number of points but placed such that a polyline close to the
      original polyline is obtained
   2. Gaussian (similar to OKS) is calculated between these approximated lines with the same number of points
4. Skeleton type:
   1. Convert the skeleton annotation into points
   2. Find the distance between these two sets of points. A set of points represents a skeleton.
   3. Here, while calculating Object Keypoint Similarity (OKS) between the points, the scale parameter is also passed,
      accounting for the spread of the skeleton. This scale is obtained from the bounding box formed around the skeleton

## Step-by-Step Guide
1. Create a new task ![new task](/images/new_task_page.jpeg)
2. Set a non zero value to `Consensus Jobs Per Regular Job` parameter under the `Advanced configuration`
   1. ![advanced config page](/images/adv_conf_page.jpeg)
   2. ![task_page_consensus](/images/task_page_consensus.jpeg)
3. Open this task, and assign workers to the consensus jobs
   1. ![consensus task](/images/consensus_task.jpeg)

>    2. ![assigned workers](/images/assigned_workers.jpeg)

Reviewer 2: This picture seems not working

Author: uf a `_` (underscore) issue

---
title: 'Consensus'
linkTitle: 'Consensus'
weight: 26
description: 'Consensus based annotation'
---

By involving multiple annotators, consensus-based annotation incorporates a range of perspectives and interpretations, which helps in minimising individual biases that might skew the annotations.

Multiple annotators reviewing and agreeing on annotations can help identify and correct errors or inconsistencies that might arise from individual biases.

For example creating bounding boxes around the face of a Basset Hound dog. One person might consider the ears also as the part of dogs face and one might not, this is due to fact that different people would have different understanding of things around them. In such cases, consensus based approach helps in reducing this bias by considering the interpretation through consensus.

| | |
|---|---|
|![basket_hound_ear_covered](/images/basket_hound_ear_covered.png)|![basket_hound_ear_not_covered](/images/basket_hound_ear_not_covered.png)|

See:

- [Use Cases](#use-cases)
- [Terminology:](#terminology)
- [Step by Step Guide](#step-by-step-guide)

## Use Cases

> 1. Instead of annotating the entire dataset with multiple annotators, you can use a smaller set of highly skilled annotators and rely on their consensus to create high-quality ground truth annotations. This reduces the overall cost compared to annotating every instance with multiple annotators.

Reviewer: It doesn't sound like a valid use case.

```suggestion
The basic use case is annotating the entire dataset with multiple annotators to obtain high-quality annotations in the dataset. In some cases it may work, but typically, however, this way of annotating is prohibitively expensive, as you need to annotate everything several times. There are several ways how the situation can be improved.

1. You can use several highly skilled annotators and rely on their consensus to create high-quality Ground Truth annotations. Typically, the Ground Truth subset contains a small portion of the whole dataset (e.g. 2-5%), so it's relatively easy to annotate it several times. These annotations are used then to validate work of different annotators randomly, so it's very desirable to avoid single expert bias in Ground Truth, and this is one of the key benefits of the consensus annotation. This method requires extra work, but it's much lower than annotating all the dataset several times. For instance, annotating a 3% of images 5 times will result in extra 3% * (5 - 1) = 12% overhead, comparing to single GT annotations.
```
